File: sync-conversations.js
#!/usr/bin/env node
// Claude Code 대화 기록을 MongoDB Atlas에 동기화하는 스크립트
// 두 가지 모드: Hook 모드(stdin으로 단일 세션) / Cron 모드(전체 스캔)

const fs = require('fs');
const path = require('path');
const os = require('os');
const { MongoClient } = require('mongodb');

require('dotenv').config({ path: path.join(__dirname, '.env') });

const CLAUDE_PROJECTS_DIR = path.join(os.homedir(), '.claude', 'projects');
const DEVICE = os.hostname();

// 명시적으로 스킵할 타입만 필터링, 나머지는 전부 유지
const SKIP_TYPES = new Set([
  'file-history-snapshot', 'progress', 'last-prompt',
  'queue-operation'
]);

// 재시도 래퍼 — 일시적 네트워크 오류 대응
async function withRetry(fn, maxRetries = 3) {
  for (let attempt = 1; attempt <= maxRetries; attempt++) {
    try {
      return await fn();
    } catch (err) {
      const isTransient = err.message.includes('ENOTFOUND') ||
        err.message.includes('timed out') ||
        err.message.includes('ECONNRESET') ||
        err.message.includes('socket disconnected');
      if (!isTransient || attempt === maxRetries) throw err;
      // 지수 백오프: 1초, 2초, 4초
      await new Promise(r => setTimeout(r, 1000 * Math.pow(2, attempt - 1)));
    }
  }
}

// JSONL 파일 파싱 — 필터링된 메시지만 반환
function parseJSONL(filePath) {
  const content = fs.readFileSync(filePath, 'utf8');
  const lines = content.trim().split('\n').filter(Boolean);
  const rawLineCount = lines.length;

  let sessionId = null;
  let sessionName = null;
  let sessionDate = null;
  let project = null;
  const messages = [];

  for (const line of lines) {
    let obj;
    try {
      obj = JSON.parse(line);
    } catch {
      continue;
    }

    const type = obj.type;

    if (!sessionId && obj.sessionId) {
      sessionId = obj.sessionId;
    }
    if (obj.type === 'custom-title' && obj.customTitle) {
      sessionName = obj.customTitle;
    }
    if (!sessionDate && obj.timestamp) {
      sessionDate = new Date(obj.timestamp);
    }
    if (!project && obj.cwd) {
      project = obj.cwd;
    }

    if (SKIP_TYPES.has(type)) continue;

    const msg = obj.message;

    // local-command-caveat 시스템 래퍼 필터링 (AI에게 보내는 무시 지시문)
    if (type === 'user' && typeof msg?.content === 'string' && msg.content.includes('<local-command-caveat>')) {
      continue;
    }

    messages.push({
      type,
      role: msg?.role,
      content: msg?.content,
      timestamp: obj.timestamp,
      uuid: obj.uuid
    });
  }

  if (!sessionId) {
    sessionId = path.basename(filePath, '.jsonl');
  }
  if (!project) {
    project = path.basename(path.dirname(filePath));
  }

  return { sessionId, sessionName, project, sessionDate, rawLineCount, messages };
}

// JSONL 파일의 라인 수만 빠르게 카운트
function countLines(filePath) {
  const content = fs.readFileSync(filePath, 'utf8');
  return content.trim().split('\n').filter(Boolean).length;
}

// 모든 JSONL 파일 탐색
function findAllJSONLFiles() {
  const files = [];

  function walk(dir) {
    let entries;
    try {
      entries = fs.readdirSync(dir, { withFileTypes: true });
    } catch {
      return;
    }
    for (const entry of entries) {
      const fullPath = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        walk(fullPath);
      } else if (entry.name.endsWith('.jsonl')) {
        files.push(fullPath);
      }
    }
  }

  walk(CLAUDE_PROJECTS_DIR);
  return files;
}

// MongoDB에 세션 upsert (재시도 포함)
async function upsertSession(collection, doc) {
  return withRetry(async () => {
    const result = await collection.updateOne(
      { session_id: doc.session_id },
      {
        $set: {
          session_name: doc.session_name,
          project: doc.project,
          device: doc.device,
          session_date: doc.session_date,
          synced_at: new Date(),
          message_count: doc.message_count,
          raw_line_count: doc.raw_line_count,
          messages: doc.messages
        }
      },
      { upsert: true }
    );
    return result.upsertedCount > 0 ? 'inserted' : (result.modifiedCount > 0 ? 'updated' : 'unchanged');
  });
}

// stdin에서 hook 데이터 읽기 시도
function readStdin() {
  return new Promise((resolve) => {
    // Windows에서는 fstatSync(0)이 TTY도 isFIFO()=false, isFile()=false로 반환
    // 안전하게: 무조건 읽기 시도, 타임아웃으로 빈 stdin 감지
    let data = '';
    let resolved = false;

    const done = (result) => {
      if (resolved) return;
      resolved = true;
      process.stdin.pause();
      process.stdin.removeAllListeners();
      resolve(result);
    };

    try {
      process.stdin.setEncoding('utf8');
      process.stdin.on('data', chunk => { data += chunk; });
      process.stdin.on('end', () => done(data.trim()));
      process.stdin.on('error', () => done(''));
      process.stdin.resume();

      // 500ms 타임아웃 — stdin이 비어있으면 cron 모드
      setTimeout(() => done(data.trim()), 500);
    } catch {
      done('');
    }
  });
}

// 단일 파일 동기화
async function syncOneFile(collection, filePath) {
  const parsed = parseJSONL(filePath);
  if (!parsed.sessionId || parsed.messages.length === 0) return null;

  const doc = {
    session_id: parsed.sessionId,
    session_name: parsed.sessionName || null,
    project: parsed.project,
    device: DEVICE,
    session_date: parsed.sessionDate || new Date(),
    message_count: parsed.messages.length,
    raw_line_count: parsed.rawLineCount,
    messages: parsed.messages
  };

  // BSON 16MB 제한 체크
  const estimatedSize = JSON.stringify(doc).length;
  if (estimatedSize > 14 * 1024 * 1024) {
    console.warn(`⚠ 세션 ${parsed.sessionId} 크기 초과 (${(estimatedSize / 1024 / 1024).toFixed(1)}MB) — 메시지 잘라냄`);
    while (JSON.stringify(doc).length > 14 * 1024 * 1024 && doc.messages.length > 10) {
      doc.messages = doc.messages.slice(0, Math.floor(doc.messages.length * 0.8));
      doc.message_count = doc.messages.length;
    }
  }

  const action = await upsertSession(collection, doc);
  return { sessionId: parsed.sessionId, action, messageCount: parsed.messages.length };
}

// 전체 스캔 (cron 모드)
async function syncAll(collection) {
  const db = collection.s.db;
  const cacheCol = db.collection('file_sync_cache');
  await cacheCol.createIndex({ file_path: 1 }, { unique: true });

  const files = findAllJSONLFiles();
  console.log(`📂 ${files.length}개 JSONL 파일 발견`);

  // 파일 경로별 라인 수 캐시 조회
  const existing = await cacheCol.find({}).toArray();
  const cacheMap = new Map(existing.map(e => [e.file_path, e.line_count]));

  let inserted = 0, updated = 0, skipped = 0, errors = 0;

  for (const filePath of files) {
    try {
      const normalizedPath = filePath.replace(/\\/g, '/');
      const currentLines = countLines(filePath);

      // 라인 수가 동일하면 스킵
      if (cacheMap.get(normalizedPath) === currentLines) {
        skipped++;
        continue;
      }

      const result = await syncOneFile(collection, filePath);
      if (result) {
        if (result.action === 'inserted') inserted++;
        else if (result.action === 'updated') updated++;
        else skipped++;

        // 캐시 갱신 (재시도 포함)
        await withRetry(() => cacheCol.updateOne(
          { file_path: normalizedPath },
          { $set: { file_path: normalizedPath, line_count: currentLines, synced_at: new Date() } },
          { upsert: true }
        ));
      }
    } catch (err) {
      errors++;
      console.error(`❌ ${path.basename(filePath)}: ${err.message}`);
    }
  }

  console.log(`✅ 완료: ${inserted} 삽입, ${updated} 갱신, ${skipped} 스킵, ${errors} 오류`);
}

// Hook 모드 — stdin에서 transcript_path 파싱
async function syncFromHook(collection, stdinData) {
  let hookData;
  try {
    hookData = JSON.parse(stdinData);
  } catch {
    console.error('stdin JSON 파싱 실패, cron 모드로 전환');
    return syncAll(collection);
  }

  const transcriptPath = hookData.transcript_path || hookData.session_id;
  if (!transcriptPath) {
    console.log('transcript_path 없음, 전체 스캔 실행');
    return syncAll(collection);
  }

  let filePath = transcriptPath;
  if (!fs.existsSync(filePath)) {
    const files = findAllJSONLFiles();
    const match = files.find(f => path.basename(f, '.jsonl') === transcriptPath);
    if (!match) {
      console.error(`파일을 찾을 수 없음: ${transcriptPath}`);
      return;
    }
    filePath = match;
  }

  const result = await syncOneFile(collection, filePath);
  if (result) {
    console.log(`✅ ${result.action}: ${result.sessionId} (${result.messageCount}개 메시지)`);
  }
}

// CLI 인자 파싱
function parseArgs() {
  const args = process.argv.slice(2);
  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--file' && args[i + 1]) return { mode: 'file', filePath: args[i + 1] };
    if (args[i] === '--scan') return { mode: 'scan' };
  }
  return { mode: 'stdin' }; // 기본값: stdin에서 hook 데이터 읽기 시도
}

// 메인 실행
async function main() {
  const uri = process.env.MONGODB_URI;
  if (!uri) {
    console.error('MONGODB_URI 환경변수가 설정되지 않음');
    process.exit(1);
  }

  const client = new MongoClient(uri, {
    serverSelectionTimeoutMS: 10000,
    connectTimeoutMS: 10000,
    socketTimeoutMS: 30000,
    retryWrites: true,
    retryReads: true
  });

  try {
    await client.connect();
    const db = client.db('conversation-warehouse');
    const collection = db.collection('sessions');

    await collection.createIndex({ session_id: 1 }, { unique: true });
    await collection.createIndex({ session_date: 1 });

    const { mode, filePath } = parseArgs();

    if (mode === 'file' && filePath) {
      // --file 모드: 특정 파일만 동기화 (가장 가벼움)
      const resolved = path.resolve(filePath);
      if (!fs.existsSync(resolved)) {
        console.error(`파일 없음: ${resolved}`);
        process.exit(1);
      }
      const result = await syncOneFile(collection, resolved);
      if (result) console.log(`✅ ${result.action}: ${result.sessionId} (${result.messageCount}개 메시지)`);
    } else if (mode === 'scan') {
      // --scan 모드: 전체 스캔 (cron용)
      await syncAll(collection);
    } else {
      // stdin 모드: hook에서 JSON 읽기
      const stdinData = await readStdin();
      if (stdinData) {
        await syncFromHook(collection, stdinData);
      } else {
        await syncAll(collection);
      }
    }
  } catch (err) {
    console.error('동기화 실패:', err.message);
    process.exit(1);
  } finally {
    await client.close();
  }
}

main();

File: sync_conversations.py
#!/usr/bin/env python3
"""Sync Claude Code conversation transcripts to MongoDB Atlas.

Modes:
  --scan       Full scan of all JSONL files (cron mode)
  --file PATH  Sync a single file
  (default)    Read stdin for hook JSON, fallback to --scan
"""

import argparse
import json
import os
import platform
import subprocess
import sys

# Windows: these hooks run under pythonw.exe, which has NO console. A console
# child (git.exe) spawned from a console-less parent must CREATE its own console
# — a window that flashes on screen at every turn end. CREATE_NO_WINDOW suppresses
# it. Zero elsewhere (the flag is Windows-only).
_NO_WINDOW = getattr(subprocess, 'CREATE_NO_WINDOW', 0) if sys.platform == 'win32' else 0
from datetime import datetime, timezone
from pathlib import Path

from _shared import force_utf8_io, get_db, to_kst_iso, with_retry

CLAUDE_PROJECTS_DIR = Path.home() / '.claude' / 'projects'
DEVICE = platform.node()

# Message types to skip
SKIP_TYPES = frozenset([
    'file-history-snapshot', 'progress', 'last-prompt', 'queue-operation'
])

TOOLKIT_REPO = Path(__file__).resolve().parent

# Version provenance: the claude-toolkit commit this sync is running under, so
# every session doc records which toolkit version produced it. Computed once per
# process (identical for every file in a scan); fails soft to None outside git.
_TOOLKIT_INFO = None


def toolkit_commit_info():
    """Return {'toolkit_commit', 'toolkit_commit_date', 'toolkit_dirty'}.

    toolkit_commit = short SHA, toolkit_commit_date = committer ISO-8601 date,
    toolkit_dirty = uncommitted changes present. Any field is None / False if git
    is unavailable. Cached for the process (same for all files in one scan).
    """
    global _TOOLKIT_INFO
    if _TOOLKIT_INFO is not None:
        return _TOOLKIT_INFO

    def _git(*args):
        return subprocess.run(
            ['git', '-C', str(TOOLKIT_REPO), *args],
            capture_output=True, text=True, timeout=5,
            creationflags=_NO_WINDOW,
        ).stdout.strip()

    info = {'toolkit_commit': None, 'toolkit_commit_date': None, 'toolkit_dirty': False}
    try:
        info['toolkit_commit'] = _git('rev-parse', '--short', 'HEAD') or None
        info['toolkit_commit_date'] = _git('log', '-1', '--format=%cI') or None
        info['toolkit_dirty'] = bool(_git('status', '--porcelain'))
    except Exception:
        pass
    _TOOLKIT_INFO = info
    return info


def drop_thinking_signatures(content):
    """Strip `signature` from extended-thinking blocks on the way into Mongo.

    An extended-thinking block arrives as {type, thinking, signature}, where the
    signature is a cryptographic seal proving the reasoning was not tampered
    with when the conversation is replayed to the API. Claude Code does not
    persist the reasoning TEXT to the transcript -- measured over 1,866 blocks,
    `thinking` was empty in every single one -- so what reaches the warehouse is
    a seal over nothing. Individual seals run from 380 to 12,000+ characters.

    It is also the worst possible thing to store: cryptographic output is
    maximum-entropy by construction (a compressible signature would be a broken
    one), so it never compresses. 187 MB across the warehouse, 37% of the whole
    database, and the reason every codec converged at ~2.5x.

    The BLOCK is kept, only the field is dropped: removing the block would change
    message shapes and block counts that readers may depend on.

    This only affects the copy written to MongoDB. ~/.claude/projects/**.jsonl is
    never modified -- `claude --resume` replays those to the API and the API
    checks the signatures, so stripping them there could break resumption.
    """
    if not isinstance(content, list):
        return content
    for block in content:
        if (isinstance(block, dict) and block.get('type') == 'thinking'
                and 'signature' in block):
            del block['signature']
    return content


def parse_jsonl(file_path):
    """Parse a JSONL transcript file, returning filtered session data."""
    text = Path(file_path).read_text(encoding='utf-8')
    lines = [l for l in text.strip().split('\n') if l]
    raw_line_count = len(lines)

    session_id = None
    session_name = None
    session_started_at = None
    project = None
    messages = []

    for line in lines:
        try:
            obj = json.loads(line)
        except json.JSONDecodeError:
            continue

        msg_type = obj.get('type')

        if not session_id and obj.get('sessionId'):
            session_id = obj['sessionId']
        if msg_type == 'custom-title' and obj.get('customTitle'):
            session_name = obj['customTitle']
        if not session_started_at and obj.get('timestamp'):
            session_started_at = datetime.fromisoformat(obj['timestamp'].replace('Z', '+00:00'))
        if not project and obj.get('cwd'):
            project = obj['cwd']

        if msg_type in SKIP_TYPES:
            continue

        msg = obj.get('message')

        # Filter local-command-caveat system wrappers
        if (msg_type == 'user' and isinstance(msg, dict)
                and isinstance(msg.get('content'), str)
                and '<local-command-caveat>' in msg['content']):
            continue

        ts = obj.get('timestamp')
        messages.append({
            'type': msg_type,
            'role': msg.get('role') if isinstance(msg, dict) else None,
            'content': drop_thinking_signatures(
                msg.get('content') if isinstance(msg, dict) else None),
            'timestamp': ts,
            'timestamp_kst': to_kst_iso(ts),
            'uuid': obj.get('uuid'),
            'parentUuid': obj.get('parentUuid'),
        })

    if not session_id:
        session_id = Path(file_path).stem
    if not project:
        project = Path(file_path).parent.name

    return {
        'session_id': session_id,
        'session_name': session_name,
        'project': project,
        'session_started_at': session_started_at,
        'raw_line_count': raw_line_count,
        'messages': messages,
    }


def count_lines(file_path):
    """Fast line count for a JSONL file."""
    text = Path(file_path).read_text(encoding='utf-8')
    return len([l for l in text.strip().split('\n') if l])


def find_all_jsonl_files():
    """Recursively find all .jsonl files under CLAUDE_PROJECTS_DIR."""
    if not CLAUDE_PROJECTS_DIR.exists():
        return []
    return list(CLAUDE_PROJECTS_DIR.rglob('*.jsonl'))


def upsert_session(collection, doc):
    """Upsert a session document with retry."""
    def _do():
        now_utc = datetime.now(timezone.utc)
        result = collection.update_one(
            {'session_id': doc['session_id']},
            {'$set': {
                'session_name': doc['session_name'],
                'project': doc['project'],
                'device': doc['device'],
                'session_started_at': doc['session_started_at'],
                'session_started_at_kst': to_kst_iso(doc['session_started_at']),
                'last_synced_at': now_utc,
                'last_synced_at_kst': to_kst_iso(now_utc),
                'message_count': doc['message_count'],
                'raw_line_count': doc['raw_line_count'],
                'messages': doc['messages'],
                'truncated': doc.get('truncated', False),
                'truncated_dropped_oldest': doc.get('truncated_dropped_oldest', 0),
                'toolkit_commit': doc.get('toolkit_commit'),
                'toolkit_commit_date': doc.get('toolkit_commit_date'),
                'toolkit_dirty': doc.get('toolkit_dirty', False),
            }},
            upsert=True,
        )
        if result.upserted_id:
            return 'inserted'
        return 'updated' if result.modified_count > 0 else 'unchanged'
    return with_retry(_do)


def sync_one_file(collection, file_path):
    """Sync a single JSONL file to MongoDB."""
    parsed = parse_jsonl(file_path)
    if not parsed['session_id'] or not parsed['messages']:
        return None

    doc = {
        'session_id': parsed['session_id'],
        'session_name': parsed['session_name'],
        'project': parsed['project'],
        'device': DEVICE,
        'session_started_at': parsed['session_started_at'] or datetime.now(timezone.utc),
        'message_count': len(parsed['messages']),
        'raw_line_count': parsed['raw_line_count'],
        'messages': parsed['messages'],
        **toolkit_commit_info(),
    }

    # BSON 16MB limit check (~14MB safety margin). Long-lived (e.g. proxy-trimmed)
    # sessions can exceed this. Keep the NEWEST messages and drop the oldest — the
    # warehouse is for recent-activity recall (mgo, quiz, reports), so the recent
    # tail is what matters. NOTE: was [:0.8] (keep-oldest), which silently dropped
    # the most recent work from any >14MB session; now [-0.8:] (keep-newest).
    estimated_size = len(json.dumps(doc, default=str, ensure_ascii=False).encode('utf-8'))
    doc['truncated'] = False
    doc['truncated_dropped_oldest'] = 0
    if estimated_size > 14 * 1024 * 1024:
        before = len(doc['messages'])
        while (len(json.dumps(doc, default=str, ensure_ascii=False).encode('utf-8')) > 14 * 1024 * 1024
               and len(doc['messages']) > 10):
            doc['messages'] = doc['messages'][-int(len(doc['messages']) * 0.8):]  # keep newest 80%
            doc['message_count'] = len(doc['messages'])
        doc['truncated'] = True
        doc['truncated_dropped_oldest'] = before - len(doc['messages'])
        # Loud: goes to stderr AND is persisted on the doc (query {truncated:true}).
        print(f'TRUNCATED {parsed["session_id"]}: kept newest {len(doc["messages"])} of '
              f'{before} msgs ({doc["truncated_dropped_oldest"]} oldest dropped) — '
              f'source was {estimated_size / 1024 / 1024:.1f}MB', file=sys.stderr)

    action = upsert_session(collection, doc)
    return {
        'session_id': parsed['session_id'],
        'action': action,
        'message_count': len(doc['messages']),
    }


def sync_all(collection):
    """Full scan: sync all JSONL files, using line-count cache to skip unchanged."""
    db = collection.database
    cache_col = db['file_sync_cache']
    cache_col.create_index('file_path', unique=True)

    files = find_all_jsonl_files()
    print(f'Found {len(files)} JSONL files')

    # Build cache map
    existing = list(cache_col.find({}))
    cache_map = {e['file_path']: e['line_count'] for e in existing}

    inserted = updated = skipped = errors = 0

    for fp in files:
        try:
            normalized = str(fp).replace('\\', '/')
            current_lines = count_lines(fp)

            if cache_map.get(normalized) == current_lines:
                skipped += 1
                continue

            result = sync_one_file(collection, fp)
            if result:
                if result['action'] == 'inserted':
                    inserted += 1
                elif result['action'] == 'updated':
                    updated += 1
                else:
                    skipped += 1

                # Update cache
                def _update_cache(n=normalized, cl=current_lines):
                    now_utc = datetime.now(timezone.utc)
                    return cache_col.update_one(
                        {'file_path': n},
                        {'$set': {'file_path': n, 'line_count': cl,
                                  'last_synced_at': now_utc,
                                  'last_synced_at_kst': to_kst_iso(now_utc)}},
                        upsert=True,
                    )
                with_retry(_update_cache)
        except Exception as e:
            errors += 1
            print(f'Error {fp.name}: {e}', file=sys.stderr)

    print(f'Done: {inserted} inserted, {updated} updated, {skipped} skipped, {errors} errors')


def sync_from_hook(collection, stdin_data):
    """Hook mode: parse stdin JSON for transcript_path and sync that file."""
    try:
        hook_data = json.loads(stdin_data)
    except json.JSONDecodeError:
        print('stdin JSON parse failed, falling back to scan', file=sys.stderr)
        return sync_all(collection)

    transcript_path = hook_data.get('transcript_path') or hook_data.get('session_id')
    if not transcript_path:
        print('No transcript_path, running full scan')
        return sync_all(collection)

    file_path = Path(transcript_path)
    if not file_path.exists():
        files = find_all_jsonl_files()
        match = next((f for f in files if f.stem == transcript_path), None)
        if not match:
            print(f'File not found: {transcript_path}', file=sys.stderr)
            return
        file_path = match

    result = sync_one_file(collection, file_path)
    if result:
        print(f'{result["action"]}: {result["session_id"]} ({result["message_count"]} messages)')


def read_stdin(timeout_ms=500):
    """Try to read stdin with a timeout. Returns data or empty string."""
    if sys.platform == 'win32':
        # Windows: select() doesn't work on stdin; check if piped
        if not sys.stdin.isatty():
            return sys.stdin.read().strip()
        return ''
    else:
        import select
        ready, _, _ = select.select([sys.stdin], [], [], timeout_ms / 1000)
        if ready:
            return sys.stdin.read().strip()
        return ''


def main():
    force_utf8_io()
    parser = argparse.ArgumentParser(description='Sync Claude conversations to MongoDB')
    parser.add_argument('--scan', action='store_true', help='Full scan mode')
    parser.add_argument('--file', dest='file_path', help='Sync a single file')
    args = parser.parse_args()

    client, db = get_db()
    try:
        collection = db['sessions']
        collection.create_index('session_id', unique=True)
        collection.create_index('session_started_at')

        if args.file_path:
            resolved = Path(args.file_path).resolve()
            if not resolved.exists():
                print(f'File not found: {resolved}', file=sys.stderr)
                sys.exit(1)
            result = sync_one_file(collection, resolved)
            if result:
                print(f'{result["action"]}: {result["session_id"]} ({result["message_count"]} messages)')
        elif args.scan:
            sync_all(collection)
        else:
            stdin_data = read_stdin()
            if stdin_data:
                sync_from_hook(collection, stdin_data)
            else:
                sync_all(collection)
    except Exception as e:
        print(f'Sync failed: {e}', file=sys.stderr)
        sys.exit(1)
    finally:
        client.close()


if __name__ == '__main__':
    main()
